tests/plugin_filter/util.py

```diff
@@ -0,0 +1,218 @@
+from hypothesis import strategies as st
+from packaging.specifiers import Specifier
+import json
+from sqlalchemy.sql import exists
+
+
+def filter_strategy():
+    return st.one_of(
+        st.fixed_dictionaries({"tag": st.text()}),
+        st.fixed_dictionaries({"name": st.text()}),
+        st.fixed_dictionaries(
+            {"version": st.from_regex(Specifier._regex)}
+        ),  # TODO: type PEP 0440 version specifier
+        st.fixed_dictionaries({"and": st.lists(st.deferred(filter_strategy))}),
+        st.fixed_dictionaries({"or": st.lists(st.deferred(filter_strategy))}),
+        st.fixed_dictionaries({"not": st.deferred(filter_strategy)}),
+    )
+
+
+def extract_key(key: str, filter_dict: dict):
+    if key in filter_dict:
+        yield filter_dict[key]
+    for k, v in filter_dict.items():
+        if k in ["and", "or"]:
+            for sub_filter in v:
+                extract_key(key, sub_filter)
+        if k == "not":
+            extract_key(key, v)
+
+
+def create_template(
+    tmp_db, name: str = "test_template", description: str = "test template"
+) -> str:
+    template_exists = tmp_db.session.query(
+        exists().where(UiTemplate.name == name)
+    ).scalar()
+    if template_exists:
+        return tmp_db.session.query(UiTemplate.id).first()[0]
+    template = UiTemplate(
+        name=name,
+        description=description,
+    )
+    tmp_db.session.add(template)
+    tmp_db.session.commit()
+    template_id = (
+        tmp_db.session.query(UiTemplate.id)
+        .filter(UiTemplate.name == template.name)
+        .first()[0]
+    )
+    return template_id
+
+
+def create_or_update_plugin(
+    tmp_db,
+    name: str = "test-plugin",
+    version: str = "0.0.0",
+    tags: list[PluginTag] | None = None,
+    description: str = "descr",
+) -> RAMP:
+    if not tags:
+        tags = []
+    plugin_exists = tmp_db.session.query(
+        exists().where(RAMP.name == name, RAMP.version == version)
+    ).scalar()
+    if plugin_exists:
+        plugin = (
+            tmp_db.session.query(RAMP)
+            .filter(RAMP.name == name, RAMP.version == version)
+            .first()
+        )
+        plugin.plugin_id = f"{name}@{version}"
+        plugin.name = name
+        plugin.version = version
+        plugin.tags = tags
+        plugin.description = description
+        tmp_db.session.commit()
+        plugin_id = (
+            tmp_db.session.query(RAMP.id)
+            .filter(RAMP.name == name, RAMP.version == version)
+            .first()[0]
+        )
+        return plugin_id
+    plugin = RAMP(
+        name=name,
+        description=description,
+        version=version,
+        tags=tags,
+    )
+    plugin.plugin_id = f"{name}@{version}"
+    tmp_db.session.add(plugin)
+    tmp_db.session.commit()
+    plugin_id = (
+        tmp_db.session.query(RAMP.id)
+        .filter(RAMP.name == name, RAMP.version == version)
+        .first()[0]
+    )
+    return plugin_id
+
+
+def create_plugins(
```

Comment:
With property based testing you need to carefully consider what should be randomized and what should be fixed. Randomization creates more test cases (and can find more errors) but also limits the possible assertions you can make in the test case (which can hide errors). Something like this, that creates some plugins in the DB may be better as a known list of plugins implemented as a pytest fixture for some tests.